src/Storages/ObjectStorage/IObjectIterator.h

```diff
@@ -5,8 +5,46 @@
 namespace DB
 {
 
-using ObjectInfo = RelativePathWithMetadata;
-using ObjectInfoPtr = std::shared_ptr<RelativePathWithMetadata>;
+namespace ErrorCodes
+{
+extern const int LOGICAL_ERROR;
+}
+
+struct ObjectInfo
+{
+    std::optional<DataLakeObjectMetadata> data_lake_metadata;
+
+    ObjectInfo() = default;
+
+    explicit ObjectInfo(const String & relative_path_)
+        : relative_path_with_metadata(RelativePathWithMetadata(relative_path_))
+    {
+    }
+    explicit ObjectInfo(RelativePathWithMetadata relative_path_with_metadata_)
+        : relative_path_with_metadata(relative_path_with_metadata_)
+    {
+    }
+
+    ObjectInfo(const ObjectInfo & other) = default;
+
+    virtual ~ObjectInfo() = default;
+
+    virtual std::string getFileName() const { return relative_path_with_metadata.getFileName(); }
+    virtual std::string getPath() const { return relative_path_with_metadata.relative_path; }
+    virtual bool isArchive() const { return false; }
+    virtual std::string getPathToArchive() const { throw Exception(ErrorCodes::LOGICAL_ERROR, "Not an archive"); }
+    virtual size_t fileSizeInArchive() const { throw Exception(ErrorCodes::LOGICAL_ERROR, "Not an archive"); }
+    virtual std::string getPathOrPathToArchiveIfArchive() const;
+    virtual std::optional<std::string> getFileFormat() const { return std::nullopt; }
+
+    std::optional<ObjectMetadata> getUnderlyingObjectMetadata() const { return relative_path_with_metadata.metadata; }
+    void setUnderlyingObjectMetadata(const ObjectMetadata & metadata) { relative_path_with_metadata.metadata = metadata; }
```

Comment:
`setUnderlyingObjectMetadata` sounds very confusing, may be name it just `setObjectMetadata` or `setPathMetadata` for example?